diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/SysInfoLinux.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/SysInfoLinux.java
index 8801b4a940f..30d33086516 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/SysInfoLinux.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/SysInfoLinux.java
@@ -138,7 +138,7 @@
    |   1  Tesla K80           Off  | 000083D4:00:00.0 Off |                    1 |
    | N/A   32C    P8    28W / 149W |     11MiB / 11439MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
-   |   2  Tesla K80           Off  | 00009D9C:00:00.0 Off |                    0 |
+   |   2  Tesla K80           Off  | 00009D9C:00:00.0 Off |                    2 |
    | N/A   29C    P8    25W / 149W |     12MiB / 11439MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
    |   3  Tesla K80           Off  | 0000B6D4:00:00.0 Off |                  N/A |
@@ -169,7 +169,7 @@
    +-----------------------------------------------------------------------------+
    */
   Pattern GPU_INFO_FORMAT =
-      Pattern.compile("\\s+([0-9]{1,2})\\s+[\\s\\S]*\\s+(0|1|N/A|Off)\\s+");
+      Pattern.compile("[|]\\s+([0-9]{1,2})[^|]*[|][^|]*[|]\\s+(\\d+|N/A|Off)\\s+[|]");
   Pattern GPU_MEM_FORMAT =
       Pattern.compile("([0-9]+)MiB\\s*/\\s*([0-9]+)MiB");
 
@@ -820,11 +820,16 @@ private void refreshGpuIfNeeded(boolean excludeOwnerlessUsingGpus, int gpuNotRea
               long index = Long.parseLong(mat.group(1));
               currentIndex = index;
 
-              String errCode = mat.group(2);
-              if (!errCode.equals("1")) {
+              int errCode;
+              try {
+                errCode = Integer.parseInt(mat.group(2));
+              } catch (NumberFormatException e) {
+                errCode = 0;
+              }
+              if (errCode == 0) {
                 gpuAttributeCapacity |= (1L << index);
               } else {
-                LOG.error("ignored error: gpu " + index + " ECC code is 1, will make this gpu unavailable");
+                LOG.error("GPU error: gpu " + index + " ECC code is " + mat.group(2) + ", will make this gpu unavailable");
               }
             }
           }
diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestSysInfoLinux.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestSysInfoLinux.java
index 52cc3f8f160..71f9c95cdbc 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestSysInfoLinux.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/util/TestSysInfoLinux.java
@@ -269,7 +269,7 @@ int readDiskBlockInformation(String diskName, int defSector) {
           "|   1  Tesla K80           Off  | 000083D4:00:00.0 Off |                    1 |\n" +
           "| N/A   32C    P8    28W / 149W |     11MiB / 11439MiB |      0%      Default |\n" +
           "+-------------------------------+----------------------+----------------------+\n" +
-          "|   2  Tesla K80           Off  | 00009D9C:00:00.0 Off |                    0 |\n" +
+          "|   2  Tesla K80           Off  | 00009D9C:00:00.0 Off |                    2 |\n" +
           "| N/A   29C    P8    25W / 149W |     12MiB / 11439MiB |      0%      Default |\n" +
           "+-------------------------------+----------------------+----------------------+\n" +
           "|   3  Tesla K80           Off  | 0000B6D4:00:00.0 Off |                  N/A |\n" +
@@ -605,8 +605,8 @@ private void InitialGPUTestFile()  throws IOException {
   public void parsingGPUFile() throws Exception {
 
     InitialGPUTestFile();
-    assertEquals(7, plugin.getNumGPUs(false, 0));
-    assertEquals(253, plugin.getGpuAttributeCapacity(false, 0));
+    assertEquals(6, plugin.getNumGPUs(false, 0));
+    assertEquals(249, plugin.getGpuAttributeCapacity(false, 0));
   }
 
 
